# Actividad 4.4.1 

## A (40 puntos) Funciones de evaluación:

Este punto es comparando directamente las listas de clase y de hipótesis, no utilices funciones preestablecidas que calculen las métricas o la matriz

1. Genera una función que reciba como parámetros una clase objetivo, la lista de clases reales y de clases asignadas por un modelo, imprima la matriz de confusión y retorne los valores de TP, TN, FP y FN. Esta función recibe 3 parámetros y retorna 4 valores

In [57]:
import pandas as pd


df = pd.read_csv('diabetes (1).csv')
df.head(5)  

preg  plas  pres  skin  insu  mass   pedi  age            class
0     3   158    76    36   245  31.6  0.851   28  tested_positive
1     1   117    60    23   106  33.8  0.466   27  tested_negative
2     1   107    50    19     0  28.3  0.181   29  tested_negative
3     1    90    62    12    43  27.2  0.580   24  tested_negative
4     2    88    58    26    16  28.4  0.766   22  tested_negative

In [58]:
def calcular_matriz_confusion_manual(clase_objetivo, reales, asignadas):
    # Inicializar contadores
    TP = TN = FP = FN = 0
    
    # Calcular valores de la matriz de confusión
    for r, a in zip(reales, asignadas):
        if r == clase_objetivo and a == clase_objetivo:
            TP += 1
        elif r == clase_objetivo and a != clase_objetivo:
            FN += 1
        elif r != clase_objetivo and a == clase_objetivo:
            FP += 1
        elif r != clase_objetivo and a != clase_objetivo:
            TN += 1
    
    # Imprimir la matriz de confusión
    print("Matriz de Confusión:")
    print(f"[[{TP}, {FN}]\n [{FP}, {TN}]]")
    
    return TP, TN, FP, FN

In [59]:
# variables de la clase
Clase_objetivo = 'Neg'
real = pd.Series(['Pos', 'Pos', 'Neg', 'Neg', 'Neg', 'Pos', 'Pos', 'Pos', 'Pos', 'Pos', 'Neg', 'Pos', 'Neg', 'Pos', 'Neg'])
hipotesis = ['Pos', 'Neg', 'Neg', 'Neg', 'Pos', 'Pos', 'Pos', 'Neg', 'Neg', 'Pos', 'Pos', 'Neg', 'Neg', 'Neg', 'Pos']

# Llamar a la función con las variables definidas
calcular_matriz_confusion_manual(Clase_objetivo, real, hipotesis)

Matriz de Confusión:
[[3, 3]
 [5, 4]]


(3, 4, 5, 3)

2. Genera una función que reciba los valores de TP, TN, FP y FN y retorne los valores de precision, recall y F1. Esta función recibe 4 parámetros y retorna 3 valores. Puedes probar tus funciones con los datos del ejercicio realizado en clase:

`Clase_objetivo='Neg'
real=pd.Series(['Pos','Pos','Neg','Neg','Neg','Pos','Pos','Pos','Pos','Pos','Neg','Pos','Neg','Pos','Neg'])
hipotesis=['Pos','Neg','Neg','Neg','Pos','Pos','Pos','Neg','Neg','Pos','Pos','Neg','Neg','Neg','Pos']`

In [60]:
# Calcular precision, recall y F1
def calcular_metricas(tp, fp, fn):
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0
    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
    
    return precision, recall, f1

# Obtener los valores de TP, TN, FP, FN de la función anterior
tp, tn, fp, fn = calcular_matriz_confusion_manual(Clase_objetivo, real, hipotesis)

# Probar la función con los datos obtenidos
precision, recall, f1 = calcular_metricas(tp, fp, fn)
print("Precision: %f,  Recall: %f, F1: %f" % (precision, recall, f1))
print("Precision: %f,  Recall: %f, F1: %f" % (precision, recall, f1))

Matriz de Confusión:
[[3, 3]
 [5, 4]]
Precision: 0.375000,  Recall: 0.500000, F1: 0.428571
Precision: 0.375000,  Recall: 0.500000, F1: 0.428571


## B (30 puntos) Generar modelos:

3. Genera al menos 5 modelos utilizando el conjunto de datos "diabetes" y el clasificador de árbol de decisión. Puedes cambiar el nivel de profundidad, instancias por hoja o el porcentaje de instancias para evaluación

In [61]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Preparación de los datos 
# Ponemos outcome como el target 
X = df.drop(columns=['class'])  
y = df['class']

# Datos de entrenamiento y de prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Hacemos 5 modelos de clasificación 
models = [
    DecisionTreeClassifier(max_depth=3),
    DecisionTreeClassifier(max_depth=5),
    DecisionTreeClassifier(min_samples_leaf=5),
    DecisionTreeClassifier(min_samples_split=10),
    DecisionTreeClassifier(max_depth=4, min_samples_leaf=3)
]

# Entrenar los modelos 
model_results = []
for i, model in enumerate(models):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    model_results.append((f"Model {i+1}", accuracy))
    print(f"Model {i+1} Accuracy: {accuracy:.4f}")

Model 1 Accuracy: 0.7922
Model 2 Accuracy: 0.7489
Model 3 Accuracy: 0.7576
Model 4 Accuracy: 0.7489
Model 5 Accuracy: 0.7619


## C (30 puntos) Evaluar:

5. Calcula la exactitud de cada modelo

4. Utiliza las funciones generadas en el punto anterior para evaluar cada uno de los modelos generado. Imprime las matrices y resultados de las métricas por modelo

In [62]:
# Evaluar cada modelo y mostrar resultados
for i, (model_name, accuracy) in enumerate(model_results):
    # Predecir con el modelo actual
    y_pred = models[i].predict(X_test)
    
    # Calcular la matriz de confusión
    print(f"{model_name}:")
    tp, tn, fp, fn = calcular_matriz_confusion_manual(Clase_objetivo, y_test, y_pred)
    
    # Calcular las métricas
    precision, recall, f1 = calcular_metricas(tp, fp, fn)
    print(f"Accuracy: {accuracy:.4f}, Precision: {precision:.4f}, Recall: {recall:.4f}, F1: {f1:.4f}\n")

Model 1:
Matriz de Confusión:
[[0, 0]
 [0, 231]]
Accuracy: 0.7922, Precision: 0.0000, Recall: 0.0000, F1: 0.0000

Model 2:
Matriz de Confusión:
[[0, 0]
 [0, 231]]
Accuracy: 0.7489, Precision: 0.0000, Recall: 0.0000, F1: 0.0000

Model 3:
Matriz de Confusión:
[[0, 0]
 [0, 231]]
Accuracy: 0.7576, Precision: 0.0000, Recall: 0.0000, F1: 0.0000

Model 4:
Matriz de Confusión:
[[0, 0]
 [0, 231]]
Accuracy: 0.7489, Precision: 0.0000, Recall: 0.0000, F1: 0.0000

Model 5:
Matriz de Confusión:
[[0, 0]
 [0, 231]]
Accuracy: 0.7619, Precision: 0.0000, Recall: 0.0000, F1: 0.0000



6. Redacta un párrafo donde menciones qué modelo obtuvo mayor: Exactitud, precision y recall (¿Fue el mismo?,¿Por qué?)

El modelo que obtuvo la mayor exactitud fue el Modelo 1 con un valor de 0.7922. Sin embargo, al evaluar las métricas de precisión y recall, es posible que otros modelos hayan tenido un mejor desempeño en estas métricas específicas dependiendo de cómo se distribuyeron los valores de TP, TN, FP y FN. Esto ocurre porque la exactitud mide el porcentaje total de predicciones correctas, mientras que la precisión y el recall se enfocan en el desempeño sobre clases específicas, lo que puede variar según el balance de clases en los datos y los parámetros del modelo. Por lo tanto, no necesariamente el modelo con mayor exactitud será el mismo que tenga la mejor precisión o recall, ya que estas métricas evalúan aspectos diferentes del desempeño del modelo.